lib/qmp.ml

```diff
@@ -99,126 +99,156 @@ module Event = struct
+    let eject json =
+      let device = json |> arguments |> U.member "device" |> U.to_string in
+      let force  = json |> arguments |> U.member "force"  |> function `Null -> None | force -> Some (force |> U.to_bool) in
+      (device, force)
+    in
+    let change json =
+      let device = json |> arguments |> U.member "device" |> U.to_string in
+      let target = json |> arguments |> U.member "target" |> U.to_string in
+      let arg    = json |> arguments |> U.member "arg"    |> function `Null -> None | arg -> Some (arg |> U.to_string) in
+      (device, target, arg)
+    in
+    let xen_save_devices_state json =
+      json |> arguments |> U.member "filename" |> U.to_string
+    in
+    let xen_load_devices_state json =
+      json |> arguments |> U.member "filename" |> U.to_string
+    in
+    let xen_set_global_dirty_log json =
+      json |> arguments |> U.member "enable" |> U.to_bool
+    in
+    let add_fd json =
+      json |> arguments |> U.member "fdset-id" |> U.to_int
+    in
+    let blockdev_change_medium json =
+      let device   = json |> arguments |> U.member "device"   |> U.to_string in
+      let filename = json |> arguments |> U.member "filename" |> U.to_string in
+      (device, filename)
+    in
+    let cmd = match json |> U.member "execute" |> U.to_string with
+    | "qmp_capabilities"         -> Qmp_capabilities
+    | "stop"                     -> Stop
+    | "cont"                     -> Cont
+    | "system_powerdown"         -> System_powerdown
+    | "query-commands"           -> Query_commands
+    | "query-status"             -> Query_status
+    | "query-vnc"                -> Query_vnc
+    | "query-kvm"                -> Query_kvm
+    | "query-xen-platform-pv-driver-info" -> Query_xen_platform_pv_driver_info
+    | "eject"                    -> json |> eject  |> fun (x, y) -> Eject (x, y)
+    | "change"                   -> json |> change |> fun (x, y, z) -> Change (x, y, z)
+    | "add-fd"                   -> json |> add_fd |> fun x -> Add_fd x
+    | "xen-save-devices-state"   -> json |> xen_save_devices_state   |> fun x -> Xen_save_devices_state x
+    | "xen-load-devices-state"   -> json |> xen_load_devices_state   |> fun x -> Xen_load_devices_state x
+    | "xen-set-global-dirty-log" -> json |> xen_set_global_dirty_log |> fun x -> Xen_set_global_dirty_log x
+    | "blockdev-change-medium"   -> json |> blockdev_change_medium   |> fun (x, y) -> Blockdev_change_medium (x, y)
+    | x -> Printf.sprintf "unknown command %s" x |> failwith
+    in
+    (json |> id, cmd)
+  in
+
+  let return json =
+    let status json =
+      json |> U.member "status" |> U.to_string
+    in
+    let query_vnc json =
+      let enabled = json |> U.member "enabled" |> U.to_bool in
+      let auth    = json |> U.member "auth"    |> U.to_string in
+      let family  = json |> U.member "family"  |> U.to_string in
+      let service = json |> U.member "service" |> U.to_string |> int_of_string in
```

Comment:
The json service field is a string; the original code then mapped it to an int and I kept the behaviour so that it still works with the existing type vnc. I'll write a comment in this line explaining this unusual bit.